Score + RNG

CubePresser opened the pull request:
4-generate-score-components

CubePresser committed:
Score context

4-generate-score-components

CubePresser committed:
RNG Context Provider

4-generate-score-components

Brallen commented on App.tsx:
```diff
@@ -22,13 +25,18 @@ const styles = StyleSheet.create({
-          <GameView />
-        </View>
-      </ImageBackground>
+      <RNGProvider seed="beer">
```
Nice seed

Brallen commented on src/shared/RNGProvider.tsx:
```diff
@@ -0,0 +1,39 @@
+const RNGContext = React.createContext<RNGStore>(null);
+export const RNGProvider: React.FC<RNGProviderProps> = ({ seed, children }) => {
+  // If we don't wrap `seedrandom` in a lambda like this, rng will be executed as state is set and it just becomes a number
+  const [ rng, setRng ] = useState<PRNG>(() => seedrandom(seed));
```
Under this way of doing it does the seed not change? Just the rng function? or this is just the initial state with a passed in seed on launch and then the setSeed method below forces the recalculation?

CubePresser replied:
Initial state w/ passed in seed. `setSeed` below regenerates the rng function w/ a new seed.

CubePresser replied:
`seedrandom(seed)` generates a `PRNG` object which is a random number generator _function_. When this provider first loads, we provide it with a `seed` prop so `rng` is initialized immediately. We can change `rng` by using `setSeed` which will generate a new `PRNG` based on the updated seed we provide. This update is communicated to any components using `const { rng } = useRNG()` and can cause re-renders where `rng` is a dependency.

You can find an example of this behavior in `Sandbox` where I've added a button that will use `setSeed`. If you look at the console, you can see that `rng` is updated, and that causes the `useMemo` which calculates voronoi data to recompute since `rng` (a dependency to this memo) has changed.

Brallen replied:
Ok cool. That makes more sense. The extra layer of the function was a little confusing in there.

Brallen commented on src/shared/ScoreProvider.tsx:
```diff
@@ -0,0 +1,43 @@
+  const [ score, setScore ] = useState(0);
+
+  const increment = () => {
+    setScore(score + 1);
```
Consider setting score as setScore((currScore) => currScore + 1). There's a bug that can happen if setScore is called multiple times in a short period where score isn't updated properly between calls so the number would be smaller than it should be.

See https://youtu.be/0yzoAbrjV6k?t=246 for a more thorough explanation.

CubePresser replied:
Nice one. I didn't know this! Thanks.

Brallen commented on src/shared/ScoreProvider.tsx:
```diff
@@ -0,0 +1,43 @@
+  };
+
+  const decrement = () => {
+    setScore(score - 1);
```
Same as above comment

Brallen reviewed (commented):
Couple of questions for you but I'm impressed with how fast you got this out. The most confusing thing to me is just how the seed and rng pieces are tied together.

CubePresser committed:
Use setState curr value callback

4-generate-score-components

CubePresser committed:
Initial Score feature implementation

Added layout + display components
Modified GameView

4-generate-score-components

CubePresser committed:
Apply styling to score display

4-generate-score-components

CubePresser commented:
This is ready to be reviewed for merge into main.

Brallen reviewed (approved):
lgtm. Good work!
